torch/distributions/transforms.py

```diff
@@ -0,0 +1,231 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all
+from torch.nn.functional import sigmoid
+
+__all__ = [
+    'AbsTransform',
+    'AffineTransform',
+    'ExpTransform',
+    'InverseTransform',
+    'LogprobTransform',
+    'SigmoidTransform',
+    'StickBreakingTransform',
+    'Transform',
+]
+
+
+class Transform(object):
+    """
+    Abstract class for transformations with computable inverse log
+    det jacobians. They are primarily used in
+    :class:`torch.distributions.TransformedDistribution`.
+
+    Transforms are intended to be short-lived objects. They memoize the forward
+    and inverse computations to avoid work; therefore :meth:`inverse` is
+    nearly free after calling :meth:`forward`. To clear the memoization cache,
+    delete the object and create a new object.
+
+    Derived classes should implement one or both of :meth:`_forward` or
+    :meth:`_inverse` and should implement :meth:`log_abs_det_jacobian`.
+    Derived classes may store intermediate results in the `._cache` dict.
+    """
+
+    def __init__(self):
+        self._cache = {}
+
+    def forward(self, x):

```

Comment:
forward/backward terms appearing in PyTorch are mostly from backpropagation/reverse-mode AD, and I don't think that this word is often used in context of transforms/bijections/diffeomorphisms. I think using `__call__` would be a nicer API overall. Then, you'd have:
```python
f = SomeTransform()
f_inv = f.inverse

f_inv(f(x)) == x
```
---
Another possibility would be to make `.inverse` a property and  instantiate an `InverseTransform(self)` inside it. Then `f.inverse` would be a first-class transform too, and you could do things like `f.inverse.log_abs_det_jacobian(a, b)` (if this pattern ends up being used often it might be good to shorten `inverse` to `inv`). Also `f.inv.inv is f`